File: fixtures/opencode/generate/opencode.py
"""Build fixtures/opencode/opencode.sqlite: two synthetic OpenCode sessions in
the database layout OpenCode itself writes.

The DDL is the generated schema, verbatim, from
packages/core/src/database/schema.gen.ts in anomalyco/opencode (the
`project`, `session`, `message` and `part` tables and their indexes). The
`data` JSON in `message` and `part` rows follows packages/schema/src/v1/
session.ts — `SessionV1.Info` minus `id`/`sessionID`, `SessionV1.Part` minus
`id`/`sessionID`/`messageID`, which is exactly what the projector strips
(packages/core/src/session/projector.ts, `messageData` / `partData`).
Rows are inserted out of the order OpenCode reads them in, so a reader that
trusted rowid order would produce a different chain. Content is synthetic.

    python fixtures/opencode/generate/opencode.py fixtures/opencode/opencode.sqlite
"""
import json, os, sqlite3, sys

out = sys.argv[1]
if os.path.exists(out):
    os.remove(out)

conn = sqlite3.connect(out)
conn.executescript(
    """
CREATE TABLE `project` (
  `id` text PRIMARY KEY,
  `worktree` text NOT NULL,
  `vcs` text,
  `name` text,
  `icon_url` text,
  `icon_url_override` text,
  `icon_color` text,
  `time_created` integer NOT NULL,
  `time_updated` integer NOT NULL,
  `time_initialized` integer,
  `sandboxes` text NOT NULL,
  `commands` text
);
CREATE TABLE `session` (
  `id` text PRIMARY KEY,
  `project_id` text NOT NULL,
  `workspace_id` text,
  `parent_id` text,
  `slug` text NOT NULL,
  `directory` text NOT NULL,
  `path` text,
  `title` text NOT NULL,
  `version` text NOT NULL,
  `share_url` text,
  `summary_additions` integer,
  `summary_deletions` integer,
  `summary_files` integer,
  `summary_diffs` text,
  `metadata` text,
  `cost` real DEFAULT 0 NOT NULL,
  `tokens_input` integer DEFAULT 0 NOT NULL,
  `tokens_output` integer DEFAULT 0 NOT NULL,
  `tokens_reasoning` integer DEFAULT 0 NOT NULL,
  `tokens_cache_read` integer DEFAULT 0 NOT NULL,
  `tokens_cache_write` integer DEFAULT 0 NOT NULL,
  `revert` text,
  `permission` text,
  `agent` text,
  `model` text,
  `time_created` integer NOT NULL,
  `time_updated` integer NOT NULL,
  `time_compacting` integer,
  `time_archived` integer,
  CONSTRAINT `fk_session_project_id_project_id_fk` FOREIGN KEY (`project_id`) REFERENCES `project`(`id`) ON DELETE CASCADE
);
CREATE TABLE `message` (
  `id` text PRIMARY KEY,
  `session_id` text NOT NULL,
  `time_created` integer NOT NULL,
  `time_updated` integer NOT NULL,
  `data` text NOT NULL,
  CONSTRAINT `fk_message_session_id_session_id_fk` FOREIGN KEY (`session_id`) REFERENCES `session`(`id`) ON DELETE CASCADE
);
CREATE TABLE `part` (
  `id` text PRIMARY KEY,
  `message_id` text NOT NULL,
  `session_id` text NOT NULL,
  `time_created` integer NOT NULL,
  `time_updated` integer NOT NULL,
  `data` text NOT NULL,
  CONSTRAINT `fk_part_message_id_message_id_fk` FOREIGN KEY (`message_id`) REFERENCES `message`(`id`) ON DELETE CASCADE
);
CREATE INDEX `message_session_time_created_id_idx` ON `message` (`session_id`,`time_created`,`id`);
CREATE INDEX `part_message_id_id_idx` ON `part` (`message_id`,`id`);
CREATE INDEX `part_session_idx` ON `part` (`session_id`);
CREATE INDEX `session_project_idx` ON `session` (`project_id`);
CREATE INDEX `session_workspace_idx` ON `session` (`workspace_id`);
"""
)

T0 = 1772697600000  # 2026-03-05T08:00:00.000Z
PROJECT = "prj_fixture0001"
conn.execute(
    "INSERT INTO project (id, worktree, vcs, name, time_created, time_updated, sandboxes) VALUES (?, ?, 'git', 'demo', ?, ?, '[]')",
    (PROJECT, "/home/dev/demo", T0, T0),
)

MODEL = {"providerID": "anthropic", "modelID": "claude-sonnet-5"}


def session(sid, title, t):
    conn.execute(
        "INSERT INTO session (id, project_id, slug, directory, title, version, cost, tokens_input, tokens_output, tokens_reasoning, tokens_cache_read, tokens_cache_write, agent, model, time_created, time_updated) "
        "VALUES (?, ?, ?, ?, ?, ?, 0, 0, 0, 0, 0, 0, 'build', ?, ?, ?)",
        (sid, PROJECT, title.lower().replace(" ", "-"), "/home/dev/demo", title, "1.2.3", json.dumps({"id": MODEL["modelID"], "providerID": MODEL["providerID"]}), t, t),
    )


rows_m, rows_p = [], []


def user(sid, mid, t, text, extra_parts=()):
    rows_m.append((mid, sid, t, t, json.dumps({"role": "user", "time": {"created": t}, "agent": "build", "model": MODEL})))
    rows_p.append((f"prt_{mid[4:]}_00", mid, sid, t, t, json.dumps({"type": "text", "text": text})))
    for i, p in enumerate(extra_parts, 1):
        rows_p.append((f"prt_{mid[4:]}_{i:02d}", mid, sid, t, t, json.dumps(p)))


def assistant(sid, mid, parent, t, parts, tokens, finish="stop"):
    # A part row is created when the part is: a tool part at its start, a
    # step-finish after the step's last part, markers a beat after the
    # previous one. `completed` is stamped after the last part.
    now = t
    stamped = []
    for i, p in enumerate(parts):
        time = p.get("time") or (p.get("state") or {}).get("time") or {}
        if "start" in time:
            now = time["start"]
        else:
            now += 50
        stamped.append((f"prt_{mid[4:]}_{i:02d}", mid, sid, now, now, json.dumps(p)))
        if "end" in time:
            now = time["end"]
    completed = now + 400
    rows_m.append(
        (
            mid, sid, t, completed,
            json.dumps({
                "role": "assistant", "time": {"created": t, "completed": completed}, "parentID": parent,
                "modelID": MODEL["modelID"], "providerID": MODEL["providerID"], "mode": "build", "agent": "build",
                "path": {"cwd": "/home/dev/demo", "root": "/home/dev/demo"}, "cost": 0.0042,
                "tokens": tokens, "finish": finish,
            }),
        )
    )
    rows_p.extend(stamped)


# --- session A: a read, an edit through a tool, a second turn ---------------
A = "ses_fixtureaaaa0001"
session(A, "Fix the parser test", T0)
user(A, "msg_a000", T0 + 1000, "The parser test is failing, can you look?")
assistant(
    A, "msg_a001", "msg_a000", T0 + 2000,
    [
        {"type": "step-start", "snapshot": "abc123"},
        {"type": "reasoning", "text": "Read the test first, then the parser.", "time": {"start": T0 + 2100, "end": T0 + 2400}},
        {"type": "text", "text": "Let me look at the failing test.", "time": {"start": T0 + 2500}},
        {"type": "tool", "callID": "call_read_1", "tool": "read",
         "state": {"status": "completed", "input": {"filePath": "test/parser.test.ts"}, "output": "expect(parse('1+1')).toBe(2)", "title": "test/parser.test.ts", "metadata": {}, "time": {"start": T0 + 2600, "end": T0 + 2700}}},
        {"type": "step-finish", "reason": "tool-calls", "cost": 0.0021, "tokens": {"input": 120, "output": 40, "reasoning": 12, "cache": {"read": 300, "write": 20}}},
        {"type": "step-start"},
        {"type": "tool", "callID": "call_edit_1", "tool": "edit",
         "state": {"status": "completed", "input": {"filePath": "src/parser.ts", "oldString": "a + b", "newString": "Number(a) + Number(b)"}, "output": "Edit applied", "title": "src/parser.ts", "metadata": {"diff": "-a + b\n+Number(a) + Number(b)"}, "time": {"start": T0 + 3000, "end": T0 + 3100}}},
        {"type": "patch", "hash": "def456", "files": ["src/parser.ts"]},
        {"type": "text", "text": "The parser was concatenating strings; fixed.", "time": {"start": T0 + 3200}},
        {"type": "step-finish", "reason": "stop", "cost": 0.0021, "tokens": {"input": 200, "output": 55, "reasoning": 0, "cache": {"read": 320, "write": 0}}},
    ],
    {"input": 320, "output": 95, "reasoning": 12, "cache": {"read": 620, "write": 20}},
)
user(A, "msg_a002", T0 + 9000, "Thanks. Run the tests.", extra_parts=[{"type": "text", "text": "Attached media from tool result:", "synthetic": True}])
assistant(
    A, "msg_a003", "msg_a002", T0 + 10000,
    [
        {"type": "tool", "callID": "call_bash_1", "tool": "bash",
         "state": {"status": "error", "input": {"command": "npm test"}, "error": "Command timed out", "metadata": {}, "time": {"start": T0 + 10100, "end": T0 + 40100}}},
        {"type": "tool", "callID": "call_bash_2", "tool": "bash",
         "state": {"status": "running", "input": {"command": "npm test -- --run"}, "time": {"start": T0 + 40200}}},
        {"type": "file", "mime": "text/plain", "url": "file:///tmp/out.txt"},
    ],
    {"input": 80, "output": 10, "reasoning": 0, "cache": {"read": 0, "write": 0}},
    finish="tool-calls",
)

# --- session B: no step-finish parts at all; usage lives on the message -----
B = "ses_fixturebbbb0002"
session(B, "Quick question", T0 + 100000)
user(B, "msg_b000", T0 + 101000, "What does the config flag do?")
assistant(
    B, "msg_b001", "msg_b000", T0 + 102000,
    [{"type": "text", "text": "It toggles verbose logging.", "time": {"start": T0 + 102100}}],
    {"input": 30, "output": 8, "reasoning": 0, "cache": {"read": 0, "write": 0}},
)

# Out of read order on purpose: the reader must sort as OpenCode does.
for r in reversed(rows_m):
    conn.execute("INSERT INTO message (id, session_id, time_created, time_updated, data) VALUES (?, ?, ?, ?, ?)", r)
for r in rows_p[::-1]:
    conn.execute("INSERT INTO part (id, message_id, session_id, time_created, time_updated, data) VALUES (?, ?, ?, ?, ?, ?)", r)
conn.commit()
for r in conn.execute("SELECT session_id, count(*) FROM message GROUP BY session_id"):
    print(r)
print("parts:", conn.execute("SELECT count(*) FROM part").fetchone()[0])
conn.close()

File: src/adapters/opencode.ts
/**
 * Adapter for OpenCode's session database (#64, #122).
 *
 * OpenCode (anomalyco/opencode) persists every session in one SQLite file,
 * `opencode.db` under its XDG data directory — `~/.local/share/opencode/`
 * unless `$XDG_DATA_HOME` says otherwise, `opencode-<channel>.db` on a
 * non-release channel (packages/core/src/database/database.ts `path()`,
 * packages/core/src/global.ts). Every name below is from the runtime's own
 * source, not from a running install:
 *
 *   - the tables, from the drizzle definitions in
 *     packages/core/src/session/sql.ts and the generated DDL in
 *     packages/core/src/database/schema.gen.ts: `session` (id, project_id,
 *     directory, title, version, model, tokens_*, time_created, …),
 *     `message` (id, session_id, time_created, time_updated, data) and
 *     `part` (id, message_id, session_id, time_created, time_updated, data);
 *   - what `data` holds, from packages/schema/src/v1/session.ts — a
 *     `message` row is `SessionV1.Info` minus `id`/`sessionID` (the
 *     projector strips exactly those two: packages/core/src/session/
 *     projector.ts `messageData`), a `part` row is `SessionV1.Part` minus
 *     `id`/`sessionID`/`messageID` (`partData`);
 *   - the read order, from `MessageV2.page` in packages/opencode/src/
 *     session/message-v2.ts: messages by `(time_created, id)`, parts by
 *     `id`, which is what the TUI shows and what this adapter reproduces.
 *
 * The mapping, per part type of the v1 schema:
 *   - `text` on a user message → `message.user`; on an assistant message →
 *     a text block. A part marked `synthetic` is OpenCode's own injected
 *     prompt (media attachments, for one) and is counted, not shown as the
 *     user's words; one marked `ignored` was never sent and is counted.
 *   - `reasoning` → a thinking block.
 *   - `tool` → `tool.call` (`callID`, `tool`, `state.input`) and, once the
 *     state is `completed` or `error`, `tool.result` (`state.output` or
 *     `state.error`). A `pending` or `running` state has no result yet;
 *     the call is recorded and the missing result counted.
 *   - `step-finish` → `cost`: one per model call, carrying that call's
 *     `tokens` (`input`, `output`, `cache.read`, `cache.write`). Reasoning
 *     tokens have no field in SPEC §5.9's usage and ride under `native`;
 *     the dollar `cost` OpenCode computes is a display-time figure and is
 *     counted, not stored (SPEC §5.9). An assistant message without any
 *     `step-finish` part but with `tokens` of its own gets one `cost` from
 *     those, so a session never loses its usage to a missing marker.
 *   - `patch` names files a git snapshot changed (`hash`, `files`) and holds
 *     no content; `snapshot` and `step-start` are markers; `file`, `agent`,
 *     `subtask`, `compaction` and `retry` are structure agit has no event
 *     for. Every one is counted by type.
 *
 * **No `file.diff`.** OpenCode's file changes live in git snapshots of the
 * worktree (`packages/opencode/src/snapshot`), not in the database, so no
 * content is available to hash and none is invented; `blame`, `why`,
 * `fork`, `merge` and `diff` have nothing to work with, everything else
 * does. Message timestamps are the rows' own (`time.created` on the
 * message, `time.start` on a tool or reasoning part, `time_created` on
 * the row), all epoch milliseconds.
 *
 * A database holds every session; `sessionsIn` lists them and `agit import`
 * takes each unless `--thread <id>` names one. Derived from the schema and
 * validated against a fixture built to it; not yet run against a real
 * `opencode.db` (#46 asks for one) — a real database that disagrees names
 * its unmapped parts in the import report rather than passing silently.
 */

import { createHash } from "node:crypto";
import type { DraftEvent, Json } from "../format/events.js";
import { looksLikeSqlite, rowsOf, SqliteError, SqliteFile, type SqliteValue } from "../sqlite.js";
import type { Adapter, ConvertOptions, ConvertResult } from "./adapter.js";

const ADAPTER_NAME = "opencode";
const ADAPTER_VERSION = "0.2.0";

type Rec = { [k: string]: Json };

function asRec(v: unknown): Rec | undefined {
  return v !== null && typeof v === "object" && !Array.isArray(v) ? (v as Rec) : undefined;
}

function str(v: Json | SqliteValue | undefined): string | null {
  return typeof v === "string" ? v : null;
}

function num(v: Json | undefined): number {
  return typeof v === "number" && Number.isFinite(v) ? v : 0;
}

/** Epoch milliseconds from a row or a JSON field, or null when it is not one. */
function ms(v: SqliteValue | Json | undefined): number | null {
  if (typeof v === "number" && Number.isFinite(v)) return v;
  if (typeof v === "bigint") return Number(v);
  return null;
}

const MIN_TS_MS = -62167219200000;
const MAX_TS_MS = 253402300799999;

function isoTs(v: number | null): string | null {
  if (v === null || v < MIN_TS_MS || v > MAX_TS_MS) return null;
  return new Date(v).toISOString();
}

interface SessionRow {
  id: string;
  directory: string | null;
  title: string | null;
  version: string | null;
  projectId: string | null;
  model: Rec | undefined;
  timeCreated: number | null;
}

interface MessageRow {
  id: string;
  sessionId: string;
  timeCreated: number | null;
  data: Rec;
}

interface PartRow {
  id: string;
  messageId: string;
  timeCreated: number | null;
  data: Rec;
}

/** A JSON column, or undefined when it is not one — the caller counts. */
function jsonColumn(v: SqliteValue | undefined): Rec | undefined {
  if (typeof v !== "string") return undefined;
  try {
    return asRec(JSON.parse(v));
  } catch {
    return undefined;
  }
}

interface Tables {
  sessions: Map<string, SessionRow>;
  messages: MessageRow[];
  parts: Map<string, PartRow[]>;
  unreadable: { messages: number; parts: number };
}

function readTables(bytes: Uint8Array): Tables {
  const db = new SqliteFile(bytes);
  const need = (name: string, columns: string[]) => {
    const t = db.table(name);
    if (t === undefined || !columns.every((c) => t.columns.includes(c))) {
      throw new Error(`this SQLite database has no OpenCode \`${name}\` table; not an opencode.db`);
    }
    return t;
  };
  const sessionT = need("session", ["id", "directory", "title", "version", "time_created"]);
  const messageT = need("message", ["id", "session_id", "time_created", "data"]);
  const partT = need("part", ["id", "message_id", "session_id", "time_created", "data"]);

  const sessions = new Map<string, SessionRow>();
  for (const r of rowsOf(db, sessionT)) {
    const id = str(r.id);
    if (id === null) continue;
    sessions.set(id, {
      id,
      directory: str(r.directory),
      title: str(r.title),
      version: str(r.version),
      projectId: str(r.project_id),
      model: jsonColumn(r.model),
      timeCreated: ms(r.time_created),
    });
  }
  const unreadable = { messages: 0, parts: 0 };
  const messages: MessageRow[] = [];
  for (const r of rowsOf(db, messageT)) {
    const id = str(r.id);
    const sessionId = str(r.session_id);
    const data = jsonColumn(r.data);
    if (id === null || sessionId === null) continue;
    if (data === undefined) {
      unreadable.messages++;
      continue;
    }
    messages.push({ id, sessionId, timeCreated: ms(r.time_created), data });
  }
  const parts = new Map<string, PartRow[]>();
  for (const r of rowsOf(db, partT)) {
    const id = str(r.id);
    const messageId = str(r.message_id);
    const data = jsonColumn(r.data);
    if (id === null || messageId === null) continue;
    if (data === undefined) {
      unreadable.parts++;
      continue;
    }
    const list = parts.get(messageId) ?? [];
    list.push({ id, messageId, timeCreated: ms(r.time_created), data });
    parts.set(messageId, list);
  }
  return { sessions, messages, parts, unreadable };
}

/** SPEC §1 wants a session id that is a safe directory name; OpenCode's `ses_…` ids are, but the rule is checked, not assumed. */
function sessionIdFor(nativeId: string): string {
  if (/^[A-Za-z0-9._-]+$/.test(nativeId) && nativeId !== "." && nativeId !== "..") return nativeId;
  return `opencode-${createHash("sha256").update(nativeId, "utf8").digest("hex").slice(0, 12)}`;
}

function cmp(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

export const opencodeAdapter: Adapter = {
  name: ADAPTER_NAME,
  version: ADAPTER_VERSION,

  /** Never text: OpenCode's log is a database. */
  detect(): boolean {
    return false;
  },

  convert(): ConvertResult {
    throw new Error("the OpenCode adapter reads opencode.db, not text; use convertBytes");
  },

  detectBytes(bytes: Uint8Array): boolean {
    if (!looksLikeSqlite(bytes)) return false;
    try {
      const db = new SqliteFile(bytes);
      const has = (name: string, cols: string[]) => {
        const t = db.table(name);
        return t !== undefined && cols.every((c) => t.columns.includes(c));
      };
      return (
        has("session", ["id", "directory", "version"]) &&
        has("message", ["id", "session_id", "data"]) &&
        has("part", ["id", "message_id", "data"])
      );
    } catch {
      return false;
    }
  },

  sessionsIn(bytes: Uint8Array): string[] {
    // Sessions with messages; an empty session is nothing to import.
    const t = readTables(bytes);
    const withMessages = new Set(t.messages.map((m) => m.sessionId));
    return [...t.sessions.keys()].filter((id) => withMessages.has(id)).sort(cmp);
  },

  convertBytes(bytes: Uint8Array, opts?: ConvertOptions): ConvertResult {
    const skipped: Record<string, number> = {};
    const skip = (what: string, n = 1): void => {
      skipped[what] = (skipped[what] ?? 0) + n;
    };

    let t: Tables;
    try {
      t = readTables(bytes);
    } catch (err) {
      if (err instanceof SqliteError)
        throw new Error(`cannot read this SQLite database: ${err.message}`, { cause: err });
      throw err;
    }
    const ids = opencodeAdapter.sessionsIn!(bytes);
    if (ids.length === 0) throw new Error("this opencode.db holds no session with messages");
    let sessionId: string;
    if (opts?.select !== undefined) {
      if (!ids.includes(opts.select)) {
        throw new Error(
          `no session ${JSON.stringify(opts.select)} in this database; sessions: ${ids.join(", ")}`,
        );
      }
      sessionId = opts.select;
    } else if (ids.length === 1) {
      sessionId = ids[0]!;
    } else {
      throw new Error(
        `this database holds ${ids.length} sessions; pass --thread <id> to pick one: ${ids.join(", ")}`,
      );
    }
    const session = t.sessions.get(sessionId)!;
    if (t.unreadable.messages > 0) skip("message-row-unreadable", t.unreadable.messages);
    if (t.unreadable.parts > 0) skip("part-row-unreadable", t.unreadable.parts);

    // The order OpenCode itself reads: messages by (time_created, id), parts by id.
    const messages = t.messages
      .filter((m) => m.sessionId === sessionId)
      .sort((a, b) => (a.timeCreated ?? 0) - (b.timeCreated ?? 0) || cmp(a.id, b.id));
    const drafts: DraftEvent[] = [];
    const firstTs = isoTs(session.timeCreated) ?? isoTs(messages[0]?.timeCreated ?? null);
    if (firstTs === null) throw new Error(`session ${sessionId} carries no timestamp agit can read`);
    let ts: string = firstTs;

    const sessionModel = session.model;
    drafts.push({
      ts,
      type: "session.start",
      payload: {
        runtime: "opencode",
        runtimeVersion: session.version,
        nativeSessionId: sessionId,
        cwd: session.directory,
        gitBranch: null,
        adapter: { name: ADAPTER_NAME, version: ADAPTER_VERSION },
        native: {
          ...(session.title !== null ? { title: session.title } : {}),
          ...(session.projectId !== null ? { projectId: session.projectId } : {}),
          ...(sessionModel !== undefined
            ? { model: { id: str(sessionModel.id), providerId: str(sessionModel.providerID) } }
            : {}),
        },
      },
    });

    let inherited = 0;
    const stamp = (v: number | null): string => {
      const iso = isoTs(v);
      if (iso === null) {
        inherited++;
        return ts;
      }
      ts = iso;
      return iso;
    };

    for (const m of messages) {
      const role = str(m.data.role);
      const time = asRec(m.data.time);
      const mts = stamp(ms(time?.created) ?? m.timeCreated);
      const parts = [...(t.parts.get(m.id) ?? [])].sort((a, b) => cmp(a.id, b.id));
      const native: Rec = { messageId: m.id };

      if (role === "user") {
        const texts: string[] = [];
        for (const p of parts) {
          const type = str(p.data.type);
          if (type === "text") {
            if (p.data.synthetic === true) skip("text-part-synthetic");
            else if (p.data.ignored === true) skip("text-part-ignored");
            else texts.push(str(p.data.text) ?? "");
          } else {
            skip(`part:${type ?? "(untyped)"}`);
          }
        }
        const text = texts.join("\n");
        if (text !== "") drafts.push({ ts: mts, type: "message.user", payload: { text, native } });
        continue;
      }

      if (role !== "assistant") {
        skip(`message-role:${role ?? "(absent)"}`);
        continue;
      }

      const model = str(m.data.modelID);
      const blocks: Json[] = [];
      const after: DraftEvent[] = [];
      let stepFinishes = 0;
      for (const p of parts) {
        const type = str(p.data.type);
        if (type === "text") {
          const text = str(p.data.text) ?? "";
          if (p.data.synthetic === true) skip("text-part-synthetic");
          else if (p.data.ignored === true) skip("text-part-ignored");
          else if (text !== "") blocks.push({ type: "text", text });
        } else if (type === "reasoning") {
          const text = str(p.data.text) ?? "";
          if (text !== "") blocks.push({ type: "thinking", text });
        } else if (type === "tool") {
          const state = asRec(p.data.state);
          const status = str(state?.status);
          const callId = str(p.data.callID);
          if (callId === null) skip("tool-part-without-call-id");
          const stime = asRec(state?.time);
          const cts = stamp(ms(stime?.start) ?? p.timeCreated);
          const input = asRec(state?.input) ?? {};
          after.push({
            ts: cts,
            type: "tool.call",
            payload: {
              toolUseId: callId,
              name: str(p.data.tool) ?? "unknown",
              input,
              native: { ...native, partId: p.id },
            },
          });
          if (status === "completed" || status === "error") {
            const rts = stamp(ms(stime?.end) ?? p.timeCreated);
            after.push({
              ts: rts,
              type: "tool.result",
              payload: {
                toolUseId: callId,
                isError: status === "error",
                output: status === "error" ? (str(state?.error) ?? "") : (str(state?.output) ?? ""),
                structured: null,
                native: {
                  ...native,
                  partId: p.id,
                  ...(str(state?.title) !== null ? { title: str(state?.title) } : {}),
                },
              },
            });
          } else {
            skip(`tool-part-${status ?? "without-state"}`);
          }
        } else if (type === "step-finish") {
          stepFinishes++;
          const tokens = asRec(p.data.tokens);
          const cache = asRec(tokens?.cache);
          if (typeof p.data.cost === "number") skip("cost-usd-not-stored (SPEC §5.9)");
          after.push({
            ts: stamp(p.timeCreated),
            type: "cost",
            payload: {
              model,
              usage: {
                inputTokens: num(tokens?.input),
                outputTokens: num(tokens?.output),
                cacheReadInputTokens: num(cache?.read),
                cacheCreationInputTokens: num(cache?.write),
              },
              native: {
                messageId: m.id,
                requestId: null,
                partId: p.id,
                reasoningTokens: num(tokens?.reasoning),
                ...(str(p.data.reason) !== null ? { reason: str(p.data.reason) } : {}),
              },
            },
          });
        } else {
          skip(`part:${type ?? "(untyped)"}`);
        }
      }
      if (blocks.length > 0) {
        drafts.push({
          ts: mts,
          type: "message.assistant",
          payload: {
            model,
            blocks,
            stopReason: str(m.data.finish),
            native: {
              ...native,
              ...(str(m.data.providerID) !== null ? { providerId: str(m.data.providerID) } : {}),
              ...(str(m.data.agent) !== null ? { agent: str(m.data.agent) } : {}),
            },
          },
        });
      }
      drafts.push(...after);
      // A message whose usage was never broken into steps still records it once.
      const tokens = asRec(m.data.tokens);
      if (stepFinishes === 0 && tokens !== undefined) {
        const cache = asRec(tokens.cache);
        if (typeof m.data.cost === "number") skip("cost-usd-not-stored (SPEC §5.9)");
        drafts.push({
          ts: stamp(ms(time?.completed) ?? null),
          type: "cost",
          payload: {
            model,
            usage: {
              inputTokens: num(tokens.input),
              outputTokens: num(tokens.output),
              cacheReadInputTokens: num(cache?.read),
              cacheCreationInputTokens: num(cache?.write),
            },
            native: { messageId: m.id, requestId: null, reasoningTokens: num(tokens.reasoning) },
          },
        });
      }
    }
    if (inherited > 0) skip("timestamp-inherited", inherited);

    if (!opts?.live) {
      drafts.push({ ts, type: "session.end", payload: { reason: "messages-end", synthesized: true } });
    }
    return { sessionId: sessionIdFor(sessionId), drafts, records: messages.length, skipped };
  },
};
